Cargo bounty cube incentives

cacogen opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

1. Bounty cube announces its creation over Supply with its creator and area, starting the speedy delivery bonus countdown
2. 20% bonus for cargo if you send the cube within 5 minutes of its creation
3. Using an export scanner on it in the supply shuttle awards a 10% personal handling tip when it reaches Centcom
4. Cube nags at five minutes over Supply if it's not sent and cargo bonus is lost
5. If there's a handler attached when it nags their ID card tells them it hasn't been sent
6. Cube continues to nag less and less often
7. All nags tell you the cube's area
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Incentivises cargo to send bounty cubes in a timely fashion.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: cacogen
add: Bounty cube announces creation over Supply with its creator and area
add: Bounty cube speedy delivery bonus of 20% for cargo if it's sent within five minutes
add: Bounty cube personal handling tip of 10% if you scan it with an export scanner while on the cargo shuttle (paid on sale)
add: Bounty cubes nag over Supply with their area, starting at five minutes with time increasing between each
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

cacogen committed:
Cargo bounty cube incentives

cacogen committed:
Pleases the linter

ArcaneMusic commented:
You've given bounty cubes tracking labels.
I'll admit, it's clever, I like this 👍

coiax commented on code/game/machinery/civilian_bounties.dm:
```diff
@@ -247,10 +244,75 @@
+	next_nag_time = world.time + nag_cooldown //grace period before the nagging begins
+
+/obj/item/bounty_cube/Destroy()
+	. = ..()
+	QDEL_NULL(radio)
```
Do cleanup logic specific to the type before the `. = ..()` call.

coiax commented on code/game/machinery/civilian_bounties.dm:
```diff
@@ -247,10 +244,75 @@
+	///How much the time between nags extends each nag.
+	var/nag_cooldown_multiplier = 1.25
+	///Next world tick to nag Supply listeners.
+	var/next_nag_time = 0
```
Use the COOLDOWN_DECLARE family of define helpers.

coiax commented on code/datums/components/pricetag.dm:
```diff
@@ -1,13 +1,16 @@
-	var/datum/bank_account/owner
-	var/profit_ratio = 1
+	var/default_profit_ratio = 1
+	var/list/payees = list()
```
Could you put an autodoc string `///` above this explaining what it's likely to contain

coiax commented on code/modules/cargo/export_scanner.dm:
```diff
@@ -9,13 +9,30 @@
 	w_class = WEIGHT_CLASS_SMALL
 
-/obj/item/export_scanner/afterattack(obj/O, mob/user, proximity)
+/obj/item/export_scanner/afterattack(obj/O, mob/living/carbon/user, proximity)
```
Is this an assertion you can make with `afterattack`?

coiax commented on code/game/machinery/civilian_bounties.dm:
```diff
@@ -247,10 +244,75 @@
+	if(speed_bonus)
+		. += "<span class='notice'><b>[time2text(bonus_countdown,"mm:ss")]</b> remaining until <b>[bounty_value * (speed_bonus/100)] cr</b> speedy delivery bonus lost.</span>"
+	if(handler_cut && !bounty_handler_account)
+		. += "<span class='notice'>Scan this in the cargo shuttle with an export scanner to claim the <b>[bounty_value * (handler_cut/100)] cr</b> handling tip."
```
Need to close this `<span>`

coiax reviewed (changes requested):
An interesting idea, would like to make some changes to make things cleaner and more consistent.

My main concern is that this is going to get too spammy far too quickly. I admit, having 3 bounty cubes pile up is a concern, but that means that in the space of 12 minutes, Cargo is going to get 6 seperate messages about it.

cacogen commented:
I understand. The potential spamminess is one of the concerns I have. But I'd prefer to see it in action before making assumptions about it. If it gets pulled I'll tweak the problems I see with it.

I don't know if six messages for three cubes is going to be all that much in twelve minutes of chat logs. If you send it in five minutes (and that seems reasonable barring someone taking off with it, which I don't think will happen often) then it's only the creation messages which are desirable because they tell you how to get your bonus. 

If it proves to be too spammy I'll have to figure out another way of getting the message to people with shuttle access or an alternative to using a message (which I assume would be more easily missed).

I was thinking of a console or a tablet app that has a list of cubes and their locations, but by itself that doesn't solve the problems of them being easy to ignore, undesirable to send and there being no motivation to send them quickly. The tablet app could let you track cubes using an overlay too.

cacogen committed:
Does what coiax asked

cacogen commented:
Ok so some issues

1. @coiax COOLDOWN_DECLARE seems to create a global variable and I'm not sure how to make it use a different variable for each cube
2. Value isn't split between cargo and the bounty holder, Centcom pays 130% of the value, should probably fix but this will devalue them for cargo (i.e. 70% vs. 100%, with 20% speed bonus optional in each case)
3. Cubes are trivially undervalued for reasons I don't understand (something to do with the k-elasticity I'm guessing), would like to fix for perfectionism but not sure how

Bonus

  i. Not sure if it can register multiple handlers but if so I need to fix that and it should probably not be overwriteable (though this creates drama so perhaps it should)
  ii. It sends the handler registered message after the handler has already been registered (makes it less clear if it worked so I will fix)

cacogen committed:
Merge remote-tracking branch 'origin/master' into bountyincentives

cacogen committed:
Ch-ch-ch-ch-changes

- Adds test cube (should've done this to start with)
- Changes autodocs to refer to cut vars as amounts instead of percentages
- Makes it so handling tip can only be registered by one bank account once

cacogen committed:
Merge remote-tracking branch 'origin/master' into bountyincentives

cacogen committed:
Bank card talk

cacogen commented:
I want to use the creation time variable I added here to tell the bounty holder how long the cube took to send, in addition to the name of the handler and the amount the holder got from it. Currently it's not used. Price tags can probably have custom sale messages set by their parent.

coiax commented on code/game/machinery/civilian_bounties.dm:
```diff
@@ -247,10 +244,100 @@
+			if(bounty_handler_account)
+				bounty_handler_account.bank_card_talk("\The [src] is unsent in <b>[get_area(src)]</b>.")
+		nag_cooldown = nag_cooldown * nag_cooldown_multiplier
+		COOLDOWN_START(src, next_nag_time, world.time + nag_cooldown)
```
COOLDOWN_START already uses `world.time`.
```suggestion
		COOLDOWN_START(src, next_nag_time, nag_cooldown)
```

coiax commented on code/game/machinery/civilian_bounties.dm:
```diff
@@ -247,10 +244,100 @@
+	///Multiplier for the bounty payout received by the person who claims the handling tip.
+	var/handler_tip = 0.1
+	///Time until speed bonus expires.
+	var/bonus_countdown = 5 MINUTES
```
I'm confused why this is a var, given that the speed bonus is lost when the cube first nags.

cacogen committed:
Adds spaces before and after the multiply symbol

Co-authored-by: coiax <yellowbounder@gmail.com>

cacogen commented on code/game/machinery/civilian_bounties.dm:
```diff
@@ -247,10 +244,100 @@
+		bonus_countdown = next_nag_time - world.time
+	if(COOLDOWN_FINISHED(src, next_nag_time))
+		if(!is_centcom_level(z) && !is_reserved_level(z)) //don't send message if we're on Centcom or in transit
+			radio.talk_into(src, "Unsent in <b>[get_area(src)]</b>.[speed_bonus ? " Speedy delivery bonus of <b>[bounty_value * speed_bonus]</b> credit\s lost." : ""]", RADIO_CHANNEL_SUPPLY)
```
Well, it obviously indicates emphasis in speech. Players not having that level of control doesn't mean other things that talk shouldn't. I doubt it's unprecedented in the codebase.

But this minor type of thing isn't worth arguing over. And the boldness is difficult to see in the monospaced font machines use anyway.

cacogen committed:
Removes the bolded values for emphasis in radio lines

Co-authored-by: coiax <yellowbounder@gmail.com>

cacogen commented on code/game/machinery/civilian_bounties.dm:
```diff
@@ -247,10 +244,100 @@
+	COOLDOWN_START(src, next_nag_time, bonus_countdown) //First nag after we've lost our speed bonus
+	radio.talk_into(src,"Created in [get_area(src)] by [bounty_holder] ([bounty_holder_job]). Speedy delivery bonus lost in [time2text(bonus_countdown,"mm:ss")].", RADIO_CHANNEL_SUPPLY)
+
+/obj/item/bounty_cube/test_cube
+	name = "test bounty cube"
+	desc = "It needs to be squeezed by someone with a bank account it can detect to set itself up. This will alert Supply."
```
I wasn't sure how IC to make it. I figured if it showed up in the round somehow then it's less immersion breaking this way. But sure.

cacogen committed:
Stress the DEBUG nature of the test cube

cacogen committed:
Adds spaces around a division symbol

Co-authored-by: coiax <yellowbounder@gmail.com>

cacogen committed:
Removes world.time being duplicated needlessly in a COOLDOWN_START

Co-authored-by: coiax <yellowbounder@gmail.com>

cacogen committed:
Removes creation_time and bonus_countdown vars

- Actually no one cares and I don't want to spend more time to make it tell you the details of your delivery on sale (i.e. who the handler was, the time it took)
- Removes bonus_countdown because it seems superfluous with the next_nag_time var as they're the same right now

cacogen committed:
Mistakes were made

cacogen commented:
@coiax okay well whoever you're ready

tgstation-server commented:
Your Fix/Feature pull request delta is currently below zero (-13). Maintainers may close future Feature/Tweak/Balance PRs. Fixing issues or helping to improve the codebase will raise this score.
